xla/service/gpu/loop_double_buffer_transformer.cc

```diff
@@ -0,0 +1,372 @@
+            for (auto new_pred : new_control_pred) {
+              TF_RETURN_IF_ERROR(new_pred->AddControlDependencyTo(new_instr));
+              VLOG(2) << "Adding " << new_pred->ToString()
+                      << " to control dependency of peeled instruction: "
+                      << new_instr->ToString();
+            }
+          }
+        }
+      }
+      // Replace all operands with newly cloned instructions.
+      for (auto instr : all_new_instructions) {
+        for (int64_t op_index = 0; op_index < instr->operand_count();
+             op_index++) {
+          VLOG(2) << "Processing operand " << op_index << " of "
+                  << instr->ToString();
+          HloInstruction* operand = instr->mutable_operand(op_index);
+          if (operand == input_parameter ||
+              operand->IsUserOf(input_parameter)) {
+            VLOG(2) << "Operand " << op_index << " is an input parameter.";
+            continue;
+          }
+          VLOG(2) << "Changing operand " << op_index << " of instruction "
+                  << instr->ToString() << " from " << operand->ToString()
+                  << " to " << old_to_new_map[operand]->ToString();
+          TF_RETURN_IF_ERROR(
+              instr->ReplaceOperandWith(op_index, old_to_new_map[operand]));
+        }
+      }
+
+      // Connect the peeled instructions with previous while loop input
+      // producers.
+      for (auto& instr_index_pair : loop_input_consumers) {
+        VLOG(2) << "Processing peeled instruction: "
+                << instr_index_pair.first->ToString();
+        HloInstruction* instr = old_to_new_map[instr_index_pair.first];
+        // Get the operand index of its consumer
+        for (auto consumer_info : instr_index_pair.second) {
+          int64_t op_index = consumer_info.operand_index;
+          HloInstruction* new_operand = loop_inputs[consumer_info.tuple_index];
+          VLOG(2) << "Changing operand " << op_index
+                  << " of peeled instruction " << instr->ToString() << " from "
+                  << instr->operand(op_index)->ToString() << " to "
+                  << new_operand->ToString();
+          if (instr->operand(op_index) != new_operand) {
+            TF_RETURN_IF_ERROR(
+                instr->ReplaceOperandWith(op_index, new_operand));
+          }
+        }
+      }
+      // Make a new tuple as the new root
+      for (int64_t i = 0; i < loop_output_count; i++) {
+        if (old_to_new_map.find(old_loop_roots[i]) != old_to_new_map.end()) {
+          new_loop_roots[i] = old_to_new_map[old_loop_roots[i]];
+        } else {
+          new_loop_roots[i] = loop_inputs[i];
+        }
+      }
+      TF_RETURN_IF_ERROR(parent_comp->ReplaceWithNewInstruction(
+          input_tuple, HloInstruction::CreateTuple(new_loop_roots)));
+      old_to_new_map.clear();
+      all_new_instructions.clear();
+      exact_trip_count -= 1;
+    }
+
+    for (auto old_instr : all_old_instructions) {
```

Comment:
The same comments regarding post-ordering and loop simplification that I made above apply here too.